hazelcast/src/main/java/com/hazelcast/map/impl/DefaultRecordStore.java

```diff
@@ -808,13 +809,17 @@ public boolean merge(Data key, EntryView mergingEntry, MapMergePolicy mergePolic
                 return true;
             }
             // same with the existing entry so no need to map-store etc operations.
-            if (mapServiceContext.compare(name, newValue, oldValue)) {
+            if (mapServiceContext.compare(name, newValue, oldValue)
+                    && mergingEntry.getExpirationTime() == existingEntry.getExpirationTime()) {
                 return true;
             }
             newValue = mapDataStore.add(key, newValue, now);
             onStore(record);
             updateSizeEstimator(-calculateRecordHeapCost(record));
             recordFactory.setValue(record, newValue);
+            if (newValue == mergingEntry.getValue()) {
+                mergeRecordExpiration(record, mergingEntry);
+            }
```

Comment:
Agreed, just need to move this code above the comparison to ensure that the TTL and expiration time of the record is updated even if the entries object is the same (the overwriting to outlast entries within ```Entry_TTL_Replication_Issue```test, is to verify this exact case).